### ISSUE TITLE
Decision trees

### ISSUE THREAD

**Issue by CodenameSource** [CodenameSource/libuaic#12]
Имплементиране на Decision tree модел за класификация

# Структура на модела
- [x] Структура на Decision tree модела
- [x] Структура на отделния възел от дървото
Псевдо код: 
```c
struct Node {
    size_t feature_index; // Индексът на feature-а който се проверява от възела
    double threshold;     // Стойността спрямо която възела проверява съответния feature
    struct Node *left, *right; 
    size_t class_prediction; // Класът който се връща ако възела е листо на дървото
};
```
- [x] Loss function, пресмятa IG([information gain](https://en.wikipedia.org/wiki/Information_gain_in_decision_trees))-а при сплит-а 
`Изваждаме сумата на ентропията на по-долните възли от ентропията на корена`
![image](https://github.com/CodenameSource/libuaic/assets/65411104/ad3a6796-71e9-4374-b45b-4f95abac74e0)
Формула за пресмятане на [ентропията](https://bg.wikipedia.org/wiki/%D0%95%D0%BD%D1%82%D1%80%D0%BE%D0%BF%D0%B8%D1%8F_%D0%BD%D0%B0_%D0%A8%D0%B0%D0%BD%D1%8A%D0%BD) за отделния node
- [x] Функция за оценяване на потенциалните разклонения на дървото
- [x] Функция за обхождане на дървото
- [x] Fit function
- [x] Predict function

# Ресурси
- [Implementing a Decision Tree From Scratch](https://towardsdatascience.com/implementing-a-decision-tree-from-scratch-f5358ff9c4bb)
- [Decision tree wikipedia](https://en.wikipedia.org/wiki/Decision_tree)
- [Decision tree scikitlearn](https://scikit-learn.org/stable/modules/tree.html)
